src/libsyntax/codemap.rs

```diff
@@ -1274,4 +1351,100 @@ mod tests {
+        // Creates an expansion chain containing two recursive calls
+        // root -> expA -> expA -> expB -> expB -> end
+        use ast::Name;
+
+        let root = Span { lo: BytePos(0), hi: BytePos(11), expn_id: NO_EXPANSION };
+
+        let format_root = ExpnFormat::MacroBang(Name(0u32));
+        let callee_root = NameAndSpan { format: format_root,
+                                        allow_internal_unstable: false,
+                                        span: Some(root) };
+
+        let info_a1 = ExpnInfo { call_site: root, callee: callee_root };
+        let id_a1 = cm.record_expansion(info_a1);
+        let span_a1 = Span { lo: BytePos(12), hi: BytePos(23), expn_id: id_a1 };
+
+        let format_a = ExpnFormat::MacroBang(Name(1u32));
+        let callee_a = NameAndSpan { format: format_a,
+                                      allow_internal_unstable: false,
+                                      span: Some(span_a1) };
+
+        let info_a2 = ExpnInfo { call_site: span_a1, callee: callee_a.clone() };
+        let id_a2 = cm.record_expansion(info_a2);
+        let span_a2 = Span { lo: BytePos(12), hi: BytePos(23), expn_id: id_a2 };
+
+        let info_b1 = ExpnInfo { call_site: span_a2, callee: callee_a };
+        let id_b1 = cm.record_expansion(info_b1);
+        let span_b1 = Span { lo: BytePos(25), hi: BytePos(36), expn_id: id_b1 };
+
+        let format_b = ExpnFormat::MacroBang(Name(2u32));
+        let callee_b = NameAndSpan { format: format_b,
+                                     allow_internal_unstable: false,
+                                     span: None };
+
+        let info_b2 = ExpnInfo { call_site: span_b1, callee: callee_b.clone() };
+        let id_b2 = cm.record_expansion(info_b2);
+        let span_b2 = Span { lo: BytePos(25), hi: BytePos(36), expn_id: id_b2 };
+
+        let info_end = ExpnInfo { call_site: span_b2, callee: callee_b };
+        let id_end = cm.record_expansion(info_end);
+        Span { lo: BytePos(37), hi: BytePos(48), expn_id: id_end }
+    }
+
+    #[test]
+    fn t12() {
+        // Test span_to_expanded_string collapses recursive macros and handles
+        // recursive callsite and callee expansions
+        let cm = init_code_map();
+        let end = init_expansion_chain(&cm);
+        let sstr = cm.span_to_expanded_string(end, "");
+        let res_str =
+            "blork2.rs:2:1: 2:12\nsecond line\n  CALLEE:\n  None\n  \
+             CALLSITE:\n  blork2.rs:1:1: 1:12\n  first line.\n    CALLEE:\n    \
+             blork.rs:2:1: 2:12\n    second line\n      CALLEE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n      CALLSITE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n    CALLSITE:\n    \
+             blork.rs:2:1: 2:12\n    second line\n      CALLEE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n      CALLSITE:\n      \
+             blork.rs:1:1: 1:12\n      first line.\n";
```

Comment:
I would use a raw string rather than putting the newlines in the string - it will be easier to read.